### PR TITLE
Simple client to view records of tables via Txn

### PR DESCRIPTION
**Pull request by manhdaovan**
As the issue here: https://github.com/hashicorp/go-memdb/issues/41
I want to add a simple client to view the records of tables in memdb.

The idea is I added a thing called `Explorer`, which can explore data of table with given table name, and a http server to display explored data.

- To init an `Explorer`, you just need passing an instance of `Txn` struct to `memdb.NewExplorer` method. Then you can use any http server to display records with your own implement.
- Or you can use `explorer_server.NewServer` to init built-in explorer server.
(You need [Gin](https://github.com/gin-gonic/gin) as dependency, and given assets and templates paths to Gin as well)

- Please ref to `explorer_example/main.go` for more details.

Here are some screenshots:
- List all tables:
screen shot 2019-02-18 at 17 30 25

- List records of a table:
  - In html format:
screen shot 2019-02-18 at 14 52 32

  - In json format:
screen shot 2019-02-18 at 14 52 45

### CODE REVIEW

File: explorer_server/list_tables.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
+package explorer_server
+
+import (
+	"github.com/gin-gonic/gin"
+	"github.com/manhdaovan/go-memdb"
```
Comment (manhdaovan) [inline explorer_server/list_tables.go]: TODO: need changing to `"github.com/hashicorp/go-memdb"` after merged.

File: explorer_server/server.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
+package explorer_server
+
+import (
+	"github.com/gin-gonic/gin"
+	"github.com/manhdaovan/go-memdb"
```
Comment (manhdaovan) [inline explorer_server/server.go]: TODO: need changing to `"github.com/hashicorp/go-memdb"` after merged.

File: explorer_example/main.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
+package main
+
+import (
+	"github.com/manhdaovan/go-memdb"
```
Comment (manhdaovan) [inline explorer_example/main.go]: TODO: need changing to `"github.com/hashicorp/go-memdb"` after merged.

**Comment by salaxander** [comment]
@manhdaovan @armon @dadgar et all - Any chance of revisiting this? It would be a hugely helpful feature for a project I'm working on :)

**Comment by dnephin** [comment]
Thank you @manhdaovan for all of your work on this! And thank you @salaxander for your interest!

I think this is a really cool idea. Being able to explore the tables and indexes in a memdb would be a really useful.

Unfortunately I don't think this is something we can maintain as part of `go-memdb`. It significantly expands the scope of this repo, and introduces new dependencies and technologies that we don't already use.

I'd like to propose the following as a path forward:

1. Create a new repo for this UI client and add the code to the new repo
2. Figure out what additional methods we need to export in `memdb` to make this UI possible from an separate package
3. Add a link to the new repo from the `go-memdb` readme or docs, so that others can find this tool

I'm happy to help with 2 and 3.  I believe currently `explorer.go` needs to access internals, but I wonder if we could take another approach. If someone where to pass in the full `memdb.DBSchema` to the explorer, would that provide a lot of the data that explorer needs? It is possible to call `Txn.Get` and get back all the entries, which I think would allow the explorer to do the necessary pagination. I only took a quick look, so maybe I have missed anything.

**Comment by salaxander** [comment]
@dnephin Sounds like a good path forward! I'm happy to help out in whatever capacity is needed :)

**Comment by manhdaovan** [comment]
Oh, I didnt have a look to this PR for a long time

@dnephin 
You are correct. Explorer needs to access to some internal stuff to get the data, so 2 is the best choice I think.

@salaxander 
The no 2 is already implemented [here](https://github.com/hashicorp/go-memdb/pull/121), then we can easily to access to schema from outside. So maybe you can get start to build an UI for it.

Btw, as no 2 is already implemented, then this PR is unnecessary anymore.

**Pull request closed without merging**